src/mlpack/core/optimizers/svrg/svrg_impl.hpp

```diff
@@ -0,0 +1,179 @@
+  arma::mat iterate0;
+
+  // Find the number of batches.
+  size_t numBatches = numFunctions / batchSize;
+  if (numFunctions % batchSize != 0)
+    ++numBatches; // Capture last few.
+
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+      std::numeric_limits<size_t>::max() : maxIterations;
+  for (size_t i = 0; i < actualMaxIterations; /* incrementing done manually */)
+  {
+    // Is this iteration the start of a sequence?
+    if ((currentFunction % numFunctions) == 0)
+    {
+      // Output current objective function.
+      Log::Info << "SVRG: iteration " << i << ", objective " << overallObjective
+          << "." << std::endl;
+
+      if (std::isnan(overallObjective) || std::isinf(overallObjective))
+      {
+        Log::Warn << "SVRG: converged to " << overallObjective
+            << "; terminating  with failure.  Try a smaller step size?"
+            << std::endl;
+        return overallObjective;
+      }
+
+      if (std::abs(lastObjective - overallObjective) < tolerance)
+      {
+        Log::Info << "SVRG: minimized within tolerance " << tolerance << "; "
+            << "terminating optimization." << std::endl;
+        return overallObjective;
+      }
+
+      // Reset the counter variables.
+      lastObjective = overallObjective;
+      overallObjective = 0;
+      currentFunction = 0;
+
+      if (shuffle) // Determine order of visitation.
+        function.Shuffle();
+    }
+
+    // Compute the full gradient.
+    size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+    function.Gradient(iterate, 0, fullGradient, effectiveBatchSize);
+    for (size_t f = effectiveBatchSize; f < numFunctions;
+        /* incrementing done manually */)
+    {
+      // Find the effective batch size (the last batch may be smaller).
+      effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+      function.Gradient(iterate, f, gradient, effectiveBatchSize);
+      fullGradient += gradient;
+
+      f += effectiveBatchSize;
+    }
+    fullGradient /= (double) numFunctions;
+
+    // Update the learning rate if requested by the user.
+    decayPolicy.Update(iterate, iterate0, gradient, fullGradient, numBatches,
+        stepSize);
+
+    // Store current parameter for the calculation of the variance reduced
+    // gradient.
+    iterate0 = iterate;
+
+    for (size_t f = 0; f < numFunctions; /* incrementing done manually */)
+    {
+      // Find the effective batch size (the last batch may be smaller).
+      effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+      // Calculate variance reduced gradient.
+      function.Gradient(iterate, f, gradient, effectiveBatchSize);
+      function.Gradient(iterate0, f, gradient0, effectiveBatchSize);
+
+      // Use the update policy to take a step.
+      updatePolicy.Update(iterate, fullGradient, gradient, gradient0,
+          effectiveBatchSize, stepSize);
+
+      f += effectiveBatchSize;
+    }
```

Comment:
This seems like it is looping over all functions, but I think you could take the inner part of this loop out (lines 142-148) and put them at line 155 and everything would work fine.

Reply:
What about introducing an update frequency for the inner loop.